File: backend/mcp-code-intelligence-kotlin/src/main/kotlin/com/codeintel/analyzers/complexity/counters/GoCounter.kt
/**
 * KSA-161: Go AST node counter.
 */
package com.codeintel.analyzers.complexity.counters

import com.codeintel.analyzers.complexity.BaseNodeCounter

class GoCounter : BaseNodeCounter() {
    override val language = "go"
    override val branchNodeTypes = setOf(
        "if_statement", "expression_case", "type_case",
    )
    override val loopNodeTypes = setOf("for_statement")
    override val logicalOperators = setOf("&&", "||")
    override val exceptionNodeTypes = setOf("defer_statement")
}

File: backend/src/engine/analyzers/complexity/counters/GoCounter.ts
/**
 * KSA-161: Go decision point counter.
 */

import type { SyntaxNode } from '../../../parsers/types.js';
import { BaseNodeCounter } from './BaseNodeCounter.js';

export class GoCounter extends BaseNodeCounter {
  readonly language = 'go';
  readonly branchNodeTypes = [
    'if_statement',
    'expression_case',       // switch case
    'type_case',             // type switch case
    'select_statement',
    'communication_case',
  ];
  readonly loopNodeTypes = [
    'for_statement',
  ];
  readonly logicalOperators = ['&&', '||'];
  readonly exceptionNodeTypes: string[] = []; // Go uses error returns, not exceptions

  protected isLogicalOp(node: SyntaxNode): boolean {
    if (node.type === 'binary_expression') {
      const op = node.childForFieldName('operator');
      if (op && this.logicalOperators.includes(op.text)) return true;
    }
    return false;
  }

  protected getReturnNodeTypes(): string[] {
    return ['return_statement'];
  }
}
